Source/Private/ASTextNode2.mm

```diff
@@ -0,0 +1,1122 @@
+{
+  _textContainer.exclusionPaths = exclusionPaths;
+  
+  [self setNeedsLayout];
+  [self setNeedsDisplay];
+}
+
+- (NSArray *)exclusionPaths
+{
+  return _textContainer.exclusionPaths;
+}
+
+- (void)prepareAttributedStringForDrawing:(NSMutableAttributedString *)attributedString
+{
+  ASDN::MutexLocker lock(__instanceLock__);
+ 
+  // Apply paragraph style if needed
+  [attributedString enumerateAttribute:NSParagraphStyleAttributeName inRange:NSMakeRange(0, attributedString.length) options:kNilOptions usingBlock:^(NSParagraphStyle *style, NSRange range, BOOL * _Nonnull stop) {
+    if (style == nil || style.lineBreakMode == _truncationMode) {
+      return;
+    }
+    
+    NSMutableParagraphStyle *paragraphStyle = [style mutableCopy] ?: [[NSMutableParagraphStyle alloc] init];
+    paragraphStyle.lineBreakMode = _truncationMode;
+    [attributedString addAttribute:NSParagraphStyleAttributeName value:paragraphStyle range:range];
+  }];
+  
+  // Apply background color if needed
+  UIColor *backgroundColor = self.backgroundColor;
+  if (CGColorGetAlpha(backgroundColor.CGColor) > 0) {
+    [attributedString addAttribute:NSBackgroundColorAttributeName value:backgroundColor range:NSMakeRange(0, attributedString.length)];
+  }
+  
+  // Apply shadow if needed
+  if (_shadowOpacity > 0 && (_shadowRadius != 0 || !CGSizeEqualToSize(_shadowOffset, CGSizeZero)) && CGColorGetAlpha(_shadowColor) > 0) {
+    NSShadow *shadow = [[NSShadow alloc] init];
+    if (_shadowOpacity != 1) {
+      shadow.shadowColor = [UIColor colorWithCGColor:CGColorCreateCopyWithAlpha(_shadowColor, _shadowOpacity * CGColorGetAlpha(_shadowColor))];
+    } else {
+      shadow.shadowColor = [UIColor colorWithCGColor:_shadowColor];
+    }
+    shadow.shadowOffset = _shadowOffset;
+    shadow.shadowBlurRadius = _shadowRadius;
+    [attributedString addAttribute:NSShadowAttributeName value:shadow range:NSMakeRange(0, attributedString.length)];
+  }
+}
+
+#pragma mark - Drawing
+
+- (NSObject *)drawParametersForAsyncLayer:(_ASDisplayLayer *)layer
+{
+  [self _ensureTruncationText];
+  ASTextContainer *copiedContainer = [_textContainer copy];
+  copiedContainer.size = self.bounds.size;
+  NSMutableAttributedString *mutableText = [self.attributedText mutableCopy] ?: [[NSMutableAttributedString alloc] init];
+  [self prepareAttributedStringForDrawing:mutableText];
+  return @{
+           @"container": copiedContainer,
+           @"text": mutableText
+           };
+}
+
+/**
+ * If it can't find a compatible layout, this method creates one.
+ */
++ (ASTextLayout *)compatibleLayoutWithContainer:(ASTextContainer *)container
+                                           text:(NSAttributedString *)text
+
+{
+  static std::mutex layoutCacheLock;
+  static NSCache<NSAttributedString *, ASTextCacheValue *> *textLayoutCache;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    textLayoutCache = [[NSCache alloc] init];
+  });
+  
+  ASTextCacheValue *cacheValue = ({
```

Comment:
What happens below when a bunch of nodes with the same text all go through this path? The first one sets an empty cache value and the rest of them get it. Then they all 'cache miss' below and layout?